tensorly/decomposition/cmtf_als.py

```diff
@@ -0,0 +1,157 @@
+    xi = lambda_ + alpha_
+
+    lambda_pred, [a_pred, B_pred, C_pred] = tl.kruskal_tensor.kruskal_normalise(kruskal_tensor_3d_pred)
+    alpha_pred, [A_pred, V_pred] = tl.kruskal_tensor.kruskal_normalise(kruskal_tensor_2d_pred)
+    xi_pred = lambda_pred + alpha_pred
+
+    # take care of permutation of normalized matrices (not mentioned in paper, but necessary)
+    perm = (A.T @ A_pred) * (B.T @ B_pred) * (C.T @ C_pred)
+    order = tl.argmax(perm, axis=1)
+    perm = perm[:, order]  # perm[r,r] now corresponds to \prod_i (a^i_r.T @ ahat^i_r) / (||(a^i_r|| ||ahat^i_r||)
+    V_pred = V_pred[:, order]
+    xi_hat = xi_pred[order]
+
+    # calculate factor match score
+    rank = A.shape[1]
+    FMS = tl.zeros(rank)
+    for r in range(rank):
+        FMS[r] = (1 - tl.abs(xi[r] - xi_hat[r]) / tl.max([xi[r], xi_hat[r]])) * tl.abs(
+            perm[r, r] * tl.dot(V[:, r], V_pred[:, r]))
+
+    return tl.min(FMS)
+
+
+def cmtf_als_for_third_order_tensor(tensor_3d, matrix, rank):
+    """
+    Calculates a coupled matrix and tensor factorization of 3rd order tensor and matrix which are coupled in first mode.
+
+    Assume you have tensor_3d = [[A, B, C]] and matrix = [[A, V]], which are coupled in 1st mode.
+    With coupled matrix and tensor factorization (CTMF), the factor matrices A, B, C for the CP decomposition of X
+    and the matrix V are found.
+    This implementation only works for a coupling in the first mode.
+
+    Solution is found via alternating least squares (ALS) as described in Figure 5 of
+    @article{acar2011all,
+      title={All-at-once optimization for coupled matrix and tensor factorizations},
+      author={Acar, Evrim and Kolda, Tamara G and Dunlavy, Daniel M},
+      journal={arXiv preprint arXiv:1105.3422},
+      year={2011}
+    }
+
+    Parameters
+    ----------
+    tensor_3d : tl.tensor or Kruskal tensor
+        3rd order tensor X = [[A, B, C]]
+    matrix : tl.tensor or Kruskal tensor
+        matrix that is coupled with tensor in first mode: Y = [[A, V]]
+    rank : int
+        rank for CP decomposition of X
+
+    Returns
+    -------
+    tensor_3d_pred, matrix_pred : Kruskal tensors
+        tensor_3d_pred = [[A,B,C]], matrix_pred = [[A,V]]
+
+    Examples
+    --------
+    A = tl.tensor([[1, 2], [3, 4]])
+    B = tl.tensor([[1, 0], [0, 2]])
+    C = tl.tensor([[2, 0], [0, 1]])
+    V = tl.tensor([[2, 0], [0, 1]])
+    R = 2
+
+    X = tl.kruskal_tensor.KruskalTensor((None, [A, B, C]))  # weights are None
+    Y = tl.kruskal_tensor.KruskalTensor((None, [A, V]))
+
+    tensor_3d_pred, matrix_pred = cmtf_als_for_third_order_tensor(X, Y, R)
+
+    """
+
+    # initialize values
+    shape_list = tensor_3d.shape
+    s = shape_list + (matrix.shape[1],)
+    A, B, C, V = random_kruskal(s, rank).factors
+
+    if isinstance(tensor_3d, tl.kruskal_tensor.KruskalTensor):
+        X = tl.kruskal_tensor.kruskal_to_tensor(tensor_3d)
+    else:
+        X = tensor_3d.copy()
```

Comment:
Why explicitly making a copy?

Reply:
In older code, sometimes I was directly assigning X = Y and when changing X, Y was also changed. This lead to an error sometimes which was difficult to find. Since them, I use the safe variant to just use a copy of the input so that I do not accidentally change it.

Reply:
But since Y is not changed, I deleted the copy here. :)